File: apps/api/internal/assistant/intent_router.go
package assistant

import "regexp"

var writeIntentDomains = map[AgentDomainId]struct{}{
	DomainContentWrite: {},
	DomainAdmin:        {},
}

var domainPatterns = []struct {
	domain  AgentDomainId
	pattern *regexp.Regexp
	weight  int
}{
	{DomainContentWrite, regexp.MustCompile(`(?i)(新建|创建|写入|删除|移除|移动|迁移|移到|挪到|跨库|move[_-]?article|重命名|归档|上传|保存|发布|撤销分享|开启分享|公开分享|改标题|更新正文|编辑文章|帮我删|帮我建)`), 3},
	{DomainAdmin, regexp.MustCompile(`(?i)(模型配置|ai\s*配置|密钥|api\s*key|公开问答|公开站|吊销|配额|开关)`), 3},
	{DomainKnowledge, regexp.MustCompile(`(?i)(知识库|文章|wiki|笔记|文档|文件|pdf|word|excel|csv)`), 2},
	{DomainSystem, regexp.MustCompile(`(多少|几个|统计|概览|状态|总数|系统|是否就绪|有没有配置)`), 2},
}

const (
	focusDomainBoost        = 4
	maxPrimaryIntentDomains = 2
)

func hasAdminDomainCandidate(domains []AgentDomainId) bool {
	for _, d := range domains {
		if d == DomainAdmin {
			return true
		}
	}
	return false
}

func withAuxiliaryDomains(domains []AgentDomainId) []AgentDomainId {
	next := append([]AgentDomainId(nil), domains...)
	hasSystem := false
	hasContentWrite := false
	for _, d := range next {
		if d == DomainSystem {
			hasSystem = true
		}
		if d == DomainContentWrite {
			hasContentWrite = true
		}
	}
	if !hasSystem {
		for _, d := range next {
			if d == DomainKnowledge {
				next = append(next, DomainSystem)
				break
			}
		}
	}
	for _, d := range next {
		if d == DomainAdmin && !hasContentWrite {
			next = append(next, DomainContentWrite)
			break
		}
	}
	return next
}

func withStickyAdminDomain(domains []AgentDomainId, recentIntentDomains []AgentDomainId) []AgentDomainId {
	hasRecentAdmin := false
	for _, d := range recentIntentDomains {
		if d == DomainAdmin {
			hasRecentAdmin = true
			break
		}
	}
	if !hasRecentAdmin {
		return domains
	}
	for _, d := range domains {
		if d == DomainAdmin {
			return domains
		}
	}
	return withAuxiliaryDomains(append(domains, DomainAdmin))
}

// RouteAssistantIntent 规则启发式意图路由（无 LLM）。
func RouteAssistantIntent(userText string, focus *AssistantFocus, recentToolNames []string, recentIntentDomains []AgentDomainId) IntentRouteResult {
	scores := map[AgentDomainId]int{}
	signals := []string{}
	bump := func(domain AgentDomainId, weight int, signal string) {
		scores[domain] += weight
		signals = append(signals, signal)
	}
	for _, p := range domainPatterns {
		if p.pattern.MatchString(userText) {
			bump(p.domain, p.weight, "text:"+string(p.domain))
		}
	}
	if focus != nil {
		if focus.KnowledgeBaseID != nil || focus.ArticleID != nil {
			bump(DomainKnowledge, focusDomainBoost, "focus:knowledge")
		}
		if focus.DocumentID != nil {
			bump(DomainKnowledge, focusDomainBoost, "focus:document")
		}
	}
	seen := map[string]struct{}{}
	for _, name := range recentToolNames {
		if _, ok := seen[name]; ok {
			continue
		}
		seen[name] = struct{}{}
		if domain := GetToolDomain(name); domain != "" {
			bump(domain, 1, "recent:"+name)
		}
	}
	type scored struct {
		domain AgentDomainId
		score  int
	}
	ranked := make([]scored, 0, len(scores))
	for d, s := range scores {
		ranked = append(ranked, scored{d, s})
	}
	for i := 0; i < len(ranked); i++ {
		for j := i + 1; j < len(ranked); j++ {
			if ranked[j].score > ranked[i].score {
				ranked[i], ranked[j] = ranked[j], ranked[i]
			}
		}
	}
	primary := make([]AgentDomainId, 0, maxPrimaryIntentDomains)
	for i := 0; i < len(ranked) && i < maxPrimaryIntentDomains; i++ {
		primary = append(primary, ranked[i].domain)
	}
	if len(primary) == 0 {
		stickyOnly := withStickyAdminDomain(append([]AgentDomainId(nil), DefaultReadDomains...), recentIntentDomains)
		conf := 0.3
		rationale := "no-signal:default-read-domains"
		if hasAdminDomainCandidate(stickyOnly) {
			for _, d := range recentIntentDomains {
				if d == DomainAdmin {
					conf = 0.7
					rationale = "no-signal:sticky-admin-domain"
					break
				}
			}
		}
		return IntentRouteResult{Domains: stickyOnly, Confidence: conf, Rationale: rationale, Source: "rules"}
	}
	domains := withStickyAdminDomain(withAuxiliaryDomains(primary), recentIntentDomains)
	conf := 0.5 + float64(len(signals))*0.1
	if conf > 0.9 {
		conf = 0.9
	}
	rationale := joinSignals(signals)
	for _, d := range recentIntentDomains {
		if d == DomainAdmin {
			for _, cd := range domains {
				if cd == DomainAdmin {
					rationale += ",sticky:admin"
					break
				}
			}
			break
		}
	}
	return IntentRouteResult{Domains: domains, Confidence: conf, Rationale: rationale, Source: "rules"}
}

func joinSignals(signals []string) string {
	if len(signals) == 0 {
		return ""
	}
	out := signals[0]
	for i := 1; i < len(signals); i++ {
		out += "," + signals[i]
	}
	return out
}

func ToIntentRoutePartData(route IntentRouteResult, status string) IntentRoutePartData {
	if status == "running" {
		return IntentRoutePartData{Status: "running", Label: "正在识别意图…"}
	}
	source := route.Source
	if source == "" {
		source = "rules"
	}
	return IntentRoutePartData{
		Status:     "done",
		Source:     source,
		Domains:    route.Domains,
		Confidence: route.Confidence,
		Rationale:  route.Rationale,
		Label:      FormatIntentRouteLabel(route.Domains, source, route.Confidence),
	}
}

func NeedsIntentLLM(route IntentRouteResult) bool {
	return route.Confidence < 0.5 || hasAdminDomainCandidate(route.Domains)
}

File: apps/web/src/server/assistant/intent-router.ts
import { DEFAULT_READ_DOMAINS, type AgentDomainId, type AssistantFocus, type IntentRouteResult } from "./domain-types"
import { getAssistantToolDomain } from "./tool-registry"
import { assistantSourceScopeFromFocus, parseAssistantSourceRef } from "@/lib/assistant-source-contract"

// 意图路由骨架（契约 4.2）：规则启发式纯打分，无 LLM 调用。
// 实现可换（规则 / 小模型），返回形状不可改。

/** 写域：装载面含危险确认相关工具 */
export const WRITE_INTENT_DOMAINS = new Set<AgentDomainId>(["content_write", "admin"])

// 写意图用更「动手」的动词；误挂靠「危险工具不暴露 + 确认协议」兜底，不再依赖 LLM 剥写域
const DOMAIN_PATTERNS: Array<{ domain: AgentDomainId; pattern: RegExp; weight: number }> = [
    {
        domain: "content_write",
        pattern: /(新建|创建|写入|删除|移除|移动|迁移|移到|挪到|跨库|move[_-]?article|重命名|归档|上传|保存|发布|撤销分享|开启分享|公开分享|改标题|更新正文|编辑文章|帮我删|帮我建)/i,
        weight: 3,
    },
    { domain: "admin", pattern: /(模型配置|ai\s*配置|密钥|api\s*key|公开问答|公开站|吊销|配额|开关)/i, weight: 3 },
    { domain: "knowledge", pattern: /(知识库|文章|wiki|笔记)/i, weight: 2 },
    { domain: "doc_library", pattern: /(文档|文件|pdf|word|excel|csv)/i, weight: 2 },
    {
        domain: "external_source",
        pattern: /(geneops|wearesellers|知无不言|微信公众号|卖家论坛|社区帖子|外部数据源|实时资料)/i,
        weight: 3,
    },
    { domain: "system", pattern: /(多少|几个|统计|概览|状态|总数|系统|是否就绪|有没有配置)/, weight: 2 },
]

const FOCUS_DOMAIN_BOOST = 4
/** 主意图域上限（辅助域另计），避免一次装载全站非 dangerous 工具 */
export const MAX_PRIMARY_INTENT_DOMAINS = 2

export function hasWriteDomainCandidate(domains: AgentDomainId[]): boolean {
    return domains.some((domain) => WRITE_INTENT_DOMAINS.has(domain))
}

export function hasAdminDomainCandidate(domains: AgentDomainId[]): boolean {
    return domains.includes("admin")
}

export function extractWriteDomains(domains: AgentDomainId[] | undefined): AgentDomainId[] {
    if (!domains?.length) return []
    return domains.filter((domain) => WRITE_INTENT_DOMAINS.has(domain))
}

/** 意图 LLM 失败时偏安全：去掉写域，保留只读面（仅用于无规则写命中的降级路径） */
export function stripWriteDomains(domains: AgentDomainId[]): AgentDomainId[] {
    const next = domains.filter((domain) => !WRITE_INTENT_DOMAINS.has(domain))
    if (next.length === 0) return [...DEFAULT_READ_DOMAINS]
    return withAuxiliaryDomains(next)
}

/**
 * admin 粘性：上一轮已激活 admin 时本轮继续挂载（admin 仍按意图按需装载）。
 * content_write 已会话常驻，不再粘性。
 */
export function withStickyAdminDomain(
    domains: AgentDomainId[],
    recentIntentDomains?: AgentDomainId[],
): AgentDomainId[] {
    if (!recentIntentDomains?.includes("admin")) return domains
    if (domains.includes("admin")) return domains
    return withAuxiliaryDomains([...domains, "admin"])
}

/** @deprecated 使用 withStickyAdminDomain；保留别名避免外部误用旧名静默失败 */
export const withStickyWriteDomains = withStickyAdminDomain

/**
 * 规则已命中 admin 时，LLM 不得剥掉（admin 保底）。
 * content_write 常驻装载，不再做写域保底。
 */
export function mergeAdminDomainFromRules(
    llmDomains: AgentDomainId[],
    rulesDomains: AgentDomainId[],
): { domains: AgentDomainId[]; kept: boolean } {
    if (!rulesDomains.includes("admin")) return { domains: llmDomains, kept: false }
    if (llmDomains.includes("admin")) return { domains: llmDomains, kept: false }
    return {
        domains: withAuxiliaryDomains(
            (["admin", ...llmDomains] as AgentDomainId[]).slice(0, MAX_PRIMARY_INTENT_DOMAINS + 2),
        ),
        kept: true,
    }
}

/** @deprecated 使用 mergeAdminDomainFromRules */
export const mergeWriteDomainsFromRules = mergeAdminDomainFromRules

export async function routeAssistantIntent(input: {
    userText: string
    focus: AssistantFocus | null
    recentToolNames: string[]
    /** 上一轮 run 的意图域；用于写域粘性 */
    recentIntentDomains?: AgentDomainId[]
}): Promise<IntentRouteResult> {
    const scores = new Map<AgentDomainId, number>()
    const signals: string[] = []
    const bump = (domain: AgentDomainId, weight: number, signal: string) => {
        scores.set(domain, (scores.get(domain) ?? 0) + weight)
        signals.push(signal)
    }

    for (const { domain, pattern, weight } of DOMAIN_PATTERNS) {
        if (pattern.test(input.userText)) bump(domain, weight, `text:${domain}`)
    }

    const focus = input.focus
    const focusDomains = new Set<AgentDomainId>()
    if (focus?.knowledgeBaseId != null || focus?.articleId != null) {
        bump("knowledge", FOCUS_DOMAIN_BOOST, "focus:knowledge")
        focusDomains.add("knowledge")
    }
    if (focus?.libraryId != null || focus?.documentId != null) {
        bump("doc_library", FOCUS_DOMAIN_BOOST, "focus:doc_library")
        focusDomains.add("doc_library")
    }
    if (focus?.sourceScope) {
        const scope = assistantSourceScopeFromFocus(focus)
        if (scope.mode === "all" || scope.mode === "local") {
            focusDomains.add("knowledge")
            focusDomains.add("doc_library")
            if (scope.mode === "all") focusDomains.add("external_source")
        } else {
            for (const ref of scope.refs) {
                const { kind } = parseAssistantSourceRef(ref)
                if (kind === "knowledge-base") focusDomains.add("knowledge")
                if (kind === "doc-library") focusDomains.add("doc_library")
                if (kind === "external-source") focusDomains.add("external_source")
            }
        }
        for (const domain of focusDomains) bump(domain, FOCUS_DOMAIN_BOOST, `scope:${domain}`)
    }

    for (const toolName of new Set(input.recentToolNames)) {
        const domain = getAssistantToolDomain(toolName)
        if (domain) bump(domain, 1, `recent:${toolName}`)
    }

    const primary = [...scores.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, MAX_PRIMARY_INTENT_DOMAINS)
        .map(([domain]) => domain)

    if (primary.length === 0) {
        // 无文本信号时仍可能粘住上一轮 admin（例如多轮管理操作中的短确认）
        const stickyOnly = withStickyAdminDomain([...DEFAULT_READ_DOMAINS], input.recentIntentDomains)
        const stickyApplied = hasAdminDomainCandidate(stickyOnly)
            && Boolean(input.recentIntentDomains?.includes("admin"))
        return {
            domains: stickyOnly,
            confidence: stickyApplied ? 0.7 : 0.3,
            rationale: stickyApplied
                ? "no-signal:sticky-admin-domain"
                : "no-signal:default-read-domains",
        }
    }

    const domains = withStickyAdminDomain(
        withAuxiliaryDomains([...new Set([...primary, ...focusDomains])]),
        input.recentIntentDomains,
    )
    const stickyAdmin = input.recentIntentDomains?.includes("admin") && domains.includes("admin")
    const rationale = [
        signals.join(","),
        ...(stickyAdmin ? ["sticky:admin"] : []),
    ].filter(Boolean).join(",")

    return {
        domains,
        confidence: Math.min(0.9, 0.5 + signals.length * 0.1),
        rationale,
    }
}

/** 资料读取域 → 补 system；admin → 补 content_write（确认工具） */
export function withAuxiliaryDomains(domains: AgentDomainId[]): AgentDomainId[] {
    let next = domains
    if ((next.includes("knowledge") || next.includes("doc_library") || next.includes("external_source"))
        && !next.includes("system")) {
        next = [...next, "system"]
    }
    if (next.includes("admin") && !next.includes("content_write")) {
        next = [...next, "content_write"]
    }
    return next
}
